ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,501 @@
+ *   and store them in its local cache. It also takes a {@link Notifier}, and subscribes to a {@link #ACCOUNT_METADATA_CHANGE_TOPIC}.
+ *   Each time when receiving a {@link #FULL_ACCOUNT_METADATA_CHANGE_MESSAGE}, it will fetch the updated full
+ *   {@link Account} metadata, and refresh its local cache. After every successful operation for updating a collection of
+ *   {@link Account}s, it publishes a {@link #FULL_ACCOUNT_METADATA_CHANGE_MESSAGE} for {@link #ACCOUNT_METADATA_CHANGE_TOPIC}
+ *   through the {@link Notifier}.
+ * </p>
+ * <p>
+ *   The full set of the {@link Account} metadata are stored in a single {@link ZNRecord}. The {@link ZNRecord}
+ *   contains {@link Account} metadata in a simple map in the following form:
+ * </p>
+ * <pre>
+ * {
+ *  {@link #ACCOUNT_METADATA_MAP_KEY} : accountMap &#60String, String&#62
+ * }
+ * </pre>
+ * <p>
+ * where {@code accountMap} is a {@link Map} from accountId String to {@link Account} JSON string.
+ * </p>
+ * <p>
+ *   Limited by {@link HelixPropertyStore}, the total size of {@link Account} data stored on a single {@link ZNRecord}
+ *   cannot exceed 1MB.
+ * </p>
+ */
+class HelixAccountService implements AccountService {
+  static final String ACCOUNT_METADATA_CHANGE_TOPIC = "account_metadata_change_topic";
+  static final String FULL_ACCOUNT_METADATA_CHANGE_MESSAGE = "full_account_metadata_change";
+  static final String ACCOUNT_METADATA_MAP_KEY = "accountMetadata";
+  static final String FULL_ACCOUNT_METADATA_PATH = "/account_metadata/full_data";
+  private static final String ZN_RECORD_ID = "full_account_metadata";
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final HelixPropertyStore<ZNRecord> helixStore;
+  private final AccountServiceMetrics accountServiceMetrics;
+  private final TopicListener<String> listener;
+  private final Notifier<String> notifier;
+  private final AtomicReference<AccountInfoMap> accountInfoMapRef = new AtomicReference<>(new AccountInfoMap());
+  private volatile boolean isOpen = false;
+  private final ReentrantLock lock = new ReentrantLock();
+
+  /**
+   * Constructor. It fetches the full account metadata set and caches locally.
+   * This call is blocking until it fetches all the {@link Account} metadata from {@link HelixPropertyStore}.
+   * @param helixStore A {@link HelixPropertyStore} used by the {@code HelixAccountService}. Cannot be {@code null}.
+   * @param accountServiceMetrics {@link AccountServiceMetrics} to report metrics. Cannot be {@code null}.
+   * @param notifier A {@link Notifier} that will be used to publish message after updating {@link Account}s, and
+   *                 listen to {@link Account} change messages. Cannot be {@code null}.
+   */
+  HelixAccountService(HelixPropertyStore<ZNRecord> helixStore, AccountServiceMetrics accountServiceMetrics,
+      Notifier<String> notifier) {
+    if (helixStore == null || accountServiceMetrics == null || notifier == null) {
```

Comment:
+1, can do `this.helixStore = Objects.requireNonNull("helixStore cannot be null")`